src/esp/gfx/RenderTarget.cpp

```diff
@@ -0,0 +1,228 @@
+const GL::Framebuffer::ColorAttachment ObjectIdBuffer =
+    GL::Framebuffer::ColorAttachment{2};
+
+struct RenderTarget::Impl {
+  Impl(WindowlessContext::ptr context, const Magnum::Vector2i& size)
+      : context_{context},
+        colorBuffer_(),
+        depthBuffer_(),
+        objectIdBuffer_(),
+        depthRenderbuffer_(),
+        framebuffer_(Magnum::NoCreate) {
+    colorBuffer_.setStorage(GL::RenderbufferFormat::SRGB8Alpha8, size);
+    depthBuffer_.setStorage(GL::RenderbufferFormat::R32F, size);
+    objectIdBuffer_.setStorage(GL::RenderbufferFormat::R32UI, size);
+    depthRenderbuffer_.setStorage(GL::RenderbufferFormat::Depth24Stencil8,
+                                  size);
+
+    framebuffer_ = GL::Framebuffer{{{}, size}};
+    framebuffer_.attachRenderbuffer(RgbaBuffer, colorBuffer_)
+        .attachRenderbuffer(DepthBuffer, depthBuffer_)
+        .attachRenderbuffer(ObjectIdBuffer, objectIdBuffer_)
+        .attachRenderbuffer(GL::Framebuffer::BufferAttachment::Depth,
+                            depthRenderbuffer_)
+        .mapForDraw({{0, RgbaBuffer}, {1, DepthBuffer}, {2, ObjectIdBuffer}});
+    CORRADE_INTERNAL_ASSERT(
+        framebuffer_.checkStatus(GL::FramebufferTarget::Draw) ==
+        GL::Framebuffer::Status::Complete);
+
+    GL::Renderer::setClearColor(Magnum::Color4{0, 0, 0, 1});
+    GL::Renderer::setClearDepth(1.0);
+  }
+
+  void renderEnter() {
+    framebuffer_.bind();
+    framebuffer_.clear(GL::FramebufferClear::Color |
+                       GL::FramebufferClear::Depth);
+  }
+
+  void renderExit() {}
+
+  void readFrameRgba(const MutableImageView2D& view) {
+    framebuffer_.mapForRead(RgbaBuffer).read(framebuffer_.viewport(), view);
+  }
+
+  void readFrameDepth(const MutableImageView2D& view) {
+    framebuffer_.mapForRead(DepthBuffer).read(framebuffer_.viewport(), view);
+  }
+
+  void readFrameObjectId(const MutableImageView2D& view) {
+    framebuffer_.mapForRead(ObjectIdBuffer).read(framebuffer_.viewport(), view);
+  }
+
+  Magnum::Vector2i framebufferSize() const {
+    return framebuffer_.viewport().size();
+  }
+
+  int gpuDeviceId() const { return this->context_->gpuDevice(); }
+
+#ifdef ESP_BUILD_GPU_GPU
+  void readFrameRgbaGPU(uint8_t* devPtr) {
+    if (colorBufferCugl_ == nullptr)
+      checkCudaErrors(cudaGraphicsGLRegisterImage(
+          &colorBufferCugl_, colorBuffer_.id(), GL_RENDERBUFFER,
+          cudaGraphicsRegisterFlagsReadOnly));
+
+    checkCudaErrors(cudaGraphicsMapResources(1, &colorBufferCugl_, 0));
+
+    cudaArray* array = nullptr;
+    checkCudaErrors(
+        cudaGraphicsSubResourceGetMappedArray(&array, colorBufferCugl_, 0, 0));
+    const int widthInBytes = framebufferSize().x() * 4 * sizeof(uint8_t);
+    checkCudaErrors(cudaMemcpy2DFromArray(devPtr, widthInBytes, array, 0, 0,
+                                          widthInBytes, framebufferSize().y(),
+                                          cudaMemcpyDeviceToDevice));
```

Comment:
Have you considered using CUDA's EGL interoperability? Using EGLImage or EGLStream, would avoid even the GPU-GPU copy. EGLStream would even potentially save on some context switching overhead. It would also be a more general solution allowing you to plug into more than just CUDA.

https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__EGL.html

Reply:
I haven't looked into this before, but I believe as long as we are using OpenGL, we absolutely need a GPU->GPU transfer as OpenGL makes no guarantees that it won't move the memory of frame buffers when you aren't actively reading (and you can't render when you are doing so).  The CUDA docs also seem to support this with the `Note that mapping of EGL resources is performed with the graphics API agnostic, resource mapping interface described in Graphics Interopability.` statement.